### PR TITLE
8323815: Source launcher should find classes with $ in names

### PR DESCRIPTION
**Pull request by sormuras**
Please review this fix to make the multi-file source-code launcher find classes with `$` in names, including names of packages, names of compliation units, and also names of enclosed types.

Tested tier1-3

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8323815](https://bugs.openjdk.org/browse/JDK-8323815): Source launcher should find classes with $ in names (**Bug** - P3)

### Reviewers
 * [Jan Lahoda](https://openjdk.org/census#jlahoda) (@lahodaj - **Reviewer**)
 * [Athijegannathan Sundararajan](https://openjdk.org/census#sundar) (@sundararajana - **Reviewer**) ⚠️ Review applies to [a8f92b90](https://git.openjdk.org/jdk/pull/17526/files/a8f92b9066fa8028234e56bf15709d19bbd17e02)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17526/head:pull/17526` \
`$ git checkout pull/17526`

Update a local copy of the PR: \
`$ git checkout pull/17526` \
`$ git pull https://git.openjdk.org/jdk.git pull/17526/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17526`

View PR using the GUI difftool: \
`$ git pr show -t 17526`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17526.diff">https://git.openjdk.org/jdk/pull/17526.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17526#issuecomment-1905425310)

### CODE REVIEW

File: test/langtools/tools/javac/launcher/MultiFileSourceLauncherTests.java [hunk 1]
```diff
@@ -86,12 +86,12 @@ record Core() {}
                     }
                     """);
         var pq = Files.createDirectories(base.resolve("p/q"));
-        Files.writeString(pq.resolve("Unit.java"),
+        Files.writeString(pq.resolve("Unit$123.java"),
```
Comment (ExE-Boss) [inline test/langtools/tools/javac/launcher/MultiFileSourceLauncherTests.java:89]: Maybe keep both tests?

File: test/langtools/tools/javac/launcher/MultiFileSourceLauncherTests.java [hunk 2]
```diff
@@ -75,7 +75,7 @@ void testLoadingOfEnclosedTypes(@TempDir Path base) throws Exception {
                     public class Hello {
                       public static void main(String... args) throws Exception {
                         System.out.println(Class.forName("World$Core"));
-                        System.out.println(Class.forName("p.q.Unit$First$Second"));
+                        System.out.println(Class.forName("p.q.Unit$123$Fir$t$$econd"));
```
Comment (lahodaj) [inline test/langtools/tools/javac/launcher/MultiFileSourceLauncherTests.java:78]: I think I would suggest two more tests:
- one that would (in a separate run) lookup `p.q.Unit$123$Fir$t` (to ensure the search is broad enough)
- one with a package name with `$`.

**Review by lahodaj — COMMENTED** [review]
I think the code is OK, but I'd suggest to add a few more tests to help ensure the behavior is kept even in case the code is modified.

**Comment by sundararajana** [comment]
LGTM

I tested the fix with jextract samples (jextract generates $ in names). All source mode + mixed .class + .java mode both work fine as expected.

**Review by sundararajana — APPROVED** [review]

File: test/langtools/tools/javac/launcher/MultiFileSourceLauncherTests.java (hunk 2 above)
Comment (sormuras) [reply test/langtools/tools/javac/launcher/MultiFileSourceLauncherTests.java:79]: Added both tests a suggested.

The first is not test in the separate run, but a separate top-level compilation unit.

File: test/langtools/tools/javac/launcher/MultiFileSourceLauncherTests.java (hunk 1 above)
Comment (sormuras) [reply test/langtools/tools/javac/launcher/MultiFileSourceLauncherTests.java:100]: Re-added the original test.

**Review by lahodaj — APPROVED** [review]
Looks reasonable.

**Pull request closed without merging**